### PR TITLE
Type hints for datatypes

### PR DESCRIPTION
**Pull request by jdavcs**
Replaces #14007 
> **From previous description:**
There are numerous occurrences of code like dataset.metadata in the code base, where dataset may refer to values of different types. This complicates reliably renaming the model.DatasetInstance.metadata property (see https://github.com/galaxyproject/galaxy/pull/13805#issuecomment-1138072044). Adding type hints is a step towards addressing this.

Types! With bug fixes along the way.

Commits have been curated: all are atomic and self-contained, with minimum cases of one commit overwriting another. The only exception are the "type-checking set_meta" commits: adding type hints to that method was nontrivial to say the least, which is why I split the process into logical steps. The set-meta commits + the subsequent 3 commits are also the only commits that will fail mypy  (each fixes one type of error caused by add types to the method definition); all other commits are mypy-compliant.

Commits should be self-explanatory. I provided details in commit descriptions as needed.

#### To do:
- [x] fix type-checking errors in tests

#### Note: method signatures of `set_meta` after refactoring:
```
(self, dataset: 'DatasetInstance', *, overwrite: bool = True, **kwd) -> None
(self, dataset: 'DatasetInstance', *, overwrite: bool = True, first_line_is_header: bool = False, **kwd) -> None
(self, dataset: 'DatasetInstance', *, overwrite: bool = True, skip: Optional[int] = None, max_data_lines: Optional[int] = 100000, max_guess_type_data_lines: Optional[int] = None, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, first_line_is_header: bool = False, metadata_tmp_files_dir: Optional[str] = None, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, metadata_tmp_files_dir: Optional[str] = None, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = 0, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = 1, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = 1, max_data_lines: Optional[int] = None, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = None, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = None, max_data_lines: Optional[int] = 5, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = None, max_data_lines: Optional[int] = 7, **kwd) -> None
(self, dataset: 'DatasetInstance', overwrite: bool = True, skip: Optional[int] = None, max_data_lines: Optional[int] = None, **kwd) -> None
```

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jdavcs** [comment]
One of the `merge` methods signatures is causing a mypy error. The error is relevant.
```
ib/galaxy/datatypes/molecules.py:553: error: Signature of "merge" incompatible with supertype "Data"  [override]
        def merge(split_files: List[str], output_file: str, extra_merge_args) -> None:
        ^
lib/galaxy/datatypes/molecules.py:553: note:      Superclass:
lib/galaxy/datatypes/molecules.py:553: note:          def merge(split_files: List[str], output_file: str) -> None
lib/galaxy/datatypes/molecules.py:553: note:      Subclass:
lib/galaxy/datatypes/molecules.py:553: note:          @staticmethod
lib/galaxy/datatypes/molecules.py:553: note:          def merge(split_files: List[str], output_file: str, extra_merge_args: Any) -> None
```

So, the superclass and all but one subclasses defined it with 2 arguments, but one method in `molecules.py` adds a third argument, which violates the Liskov Substitution principle.

There are at least 2 solutions:
a) Add `extra_merge_args: Optional[Any]` as the third argument to all other definitions of `merge` (there are 13). 
b) Drop the `extra_merge_args` argument.

As far as I can tell, this argument was never used. The only place in the code base that calls `merge` with 3 arguments is [here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/splitters/multi.py#L168) - in `jobs/splitters/multi.py`. The change was made [10 years ago](https://github.com/galaxyproject/galaxy/commit/2e9f117735cbc21f8dcbf9feb4adf34571366f82), but none of the `merge` methods at that time supported it. Then, in 2016 the molecules datatypes were added (ref #1941), at which point a `merge` method definition with 3 arguments was added. However, that method [raises](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/molecules.py#L547) a `NotImplementedError.`

Should we drop the argument, as well as the modification in `jobs/splitters/multi.py`? Or is this functionality we are likely to use? (in which case the other signatures will have to be updated) For now I'll add a type ignore. 
Ping @jmchilton, @bgruening

**Comment by bgruening** [comment]
Sorry for being late. I'm fine if you drop it.

File: lib/galaxy/datatypes/anvio.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/anvio.py:18]: A Protocol might be a good idea here, I guess this only needs to provide `file_name` ?

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
    from galaxy.util.compression_utils import FileObjType
```

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py:646]: Why not `DatasetInstance` ?

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py:334]: Why not `Datatype` ?

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py:388]: Why not `Optional[bool]` as the return type? The `False` seems quite unnecessary. Also wouldn't it be enough to annotate the parent in the class hierarchy ?

File: lib/galaxy/model/__init__.py
Comment (mvdbeek) [inline lib/galaxy/model/__init__.py]: That looks unused, drop ?

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: Is this not going through type checks ? This should've made something fail ...

File: lib/galaxy/datatypes/anvio.py
Comment (jmchilton) [reply lib/galaxy/datatypes/anvio.py:18]: Even if more is needed than file_name in a lot of cases - tracking what is needed would be great. I would love to see a second application to these datatype modules decoupled from our models and Galaxy - there is a lot of useful code and domain knowledge in here and while I've never been able to strictly separate it from the models (hence galaxy-data packaging them both) the actual interconnectedness seems like it could be pretty loose.

File: lib/galaxy/datatypes/binary.py
Comment (jmchilton) [reply lib/galaxy/datatypes/binary.py]: If you think about the type checking that happens in the packages when we import say something from structured app in a package that doesn't depend on galaxy-app (lower in the dependency tree) it doesn't fail. When I first saw people adding those imports - I thought for sure they would fail but they don't - I guess type checking imports are all optional? I imagine there is a mypy flag for this - but there are places in the project where we now depend on this behavior.

**Review by jmchilton — APPROVED** [review]

File: lib/galaxy/datatypes/binary.py
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py:334]: I'm not entirely sure what type this is. Some older typing appears to suggest it could be a list of strings (which it is not, but still). It expects the item in the list to have an attribute, `compressed_format` - and that would make it a `FilePrefix`. I couldn't easily trace this, so I decided to leave this as `Any`. Unless there's an obvious solution, this could lead down a garden path.. My vote is to tighten it in another PR.

File: lib/galaxy/datatypes/binary.py
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py:388]: Given that `sniff` explicitly returns `True` when successful, I think returning `True` OR `False` is more straightforward than `True` OR `False` OR `None`. There seems to be no reason to not return an explicit boolean value (i.e., there's no case where `None` is more appropriate than `False`) other than to not include that extra line. I'd rather keep it as bool so it's more explicit.

It is indeed enough to annotate the parent for all its subclasses to be type-checked. Again, for the sake of readability I'd keep the return type on all method definitions (so one doesn't have to go hunting for the top of the class hierarchy where the type is specified.

File: lib/galaxy/model/__init__.py
Comment (jdavcs) [reply lib/galaxy/model/__init__.py]: this one is done.

File: lib/galaxy/datatypes/binary.py
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py]: The path was incorrect too: it's `galaxy.util`, not `galaxy.datatypes.util`. Fixed.

I have no idea why it didn't fail. The docs (both mypy and python, PEP484) offer no clarification. I guess it's entirely up to mypy how to handle these imports. Apparently, you can `import gibberish from gibberish`, then do anything with `gibberish` inside an annotated function that is type-checked, and it'll pass just fine. But if you do the same for a real import (still inside the TYPE_CHECKING block), and then try using a non-existent attribute, then it'll raise an error. Same for trying `from galaxy import gibberish`. So, it seems, it ignores imports it cannot find and treats them as `Any`.

File: lib/galaxy/datatypes/binary.py
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py:646]: Somewhere in the definition of `display_data` in the parent class, there's a reference to `dataset.hid` - so that makes it an HDA (no hid on a DatasetInstance).

File: lib/galaxy/datatypes/anvio.py
Comment (jdavcs) [reply lib/galaxy/datatypes/anvio.py:18]: On decoupling from the model. We import a lot from `model.metadata` in `datatypes/metadata.py`. But other than that, it's just DatasetInstance, HDA, and one case of HDCA. Setting up Protocols for those might be not too difficult. I might try it during next week's "typing-a-thon".

**Comment by jdavcs** [comment]
Thank you for the thorough review!  
@mvdbeek I'll merge this once it's green, since @jmchilton has approved it - so that we have it in dev before next week's hackathon. If you'd like me to address any of the remaining items (I've made changes or commented on all the suggestions), I'll do that as a separate PR.

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py:646]: Remove that then ? It makes no sense, and you should be able to use this from e.g. the library.

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py:388]: I really dislike all this verbosity, if we change something we'll have to change everything now ? Most IDEs autocomplete and warn you if the return type doesn't match, so I don't think that's a good argument.

**Comment by mvdbeek** [comment]
I guess I didn't formally block this, but maybe don't rush it so hard given I had concerns here ? Like you also ignored John and my request to add a Protocol. We don't need to revert, but this is going a bit too far for my taste ?

**Comment by jdavcs** [comment]
> I guess I didn't formally block this, but maybe don't rush it so hard given I had concerns here ? Like you also ignored John and my request to add a Protocol. We don't need to revert, but this is going a bit too far for my taste ?

I never ignore reviewer comments, and I did not ignore neither John's not yours this time. Please see my response to the Protocols discussion [here](https://github.com/galaxyproject/galaxy/pull/15052#discussion_r1044858405). Furthermore, we've discussed this briefly in the past ([ref](https://github.com/galaxyproject/galaxy/pull/14007#pullrequestreview-998517602)) and agreed that while protocols are the right way to go, using forward references is not a blocker; furthermore, in this particular case they are from the same package.

File: lib/galaxy/datatypes/binary.py
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py:388]: Not everyone uses an IDE. But sure, I can remove the annotations from methods in subclasses where that doesn't affect type-checking. I'll do that as a follow-up PR.

**Pull request merged**
